### ISSUE TITLE
Weird crash and tentative fix

### ISSUE THREAD

**Issue by RJVB** [zaps166/QMPlay2#137]
[more-readable.diff.txt](https://github.com/zaps166/QMPlay2/files/1628240/more-readable.diff.txt)
I've seen a couple of very strange crashes on Mac with the latest version from git, built with clang 5.0.1 .

I didn't keep backtraces, unfortunately, but the strangeness started under `VideoFilters::clearBuffers()`  when `filtersThr.waitForFinished(true);` behaved like `filtersThr->waitForFinished(true);` with `filtersThr == NULL`.

After staring a bit at the code I noticed this is actually not so very strange because filtersThr is initialised from a dynamically allocated instance.

I didn't like that extra dereferencing and copying which just might give compilers the idea that they're dealing with a variable on the stack that can go out of scope. So I made the attached change hoping it woud help find where the nullptr came from.

Until now that hope hasn't come true: the crash doesn't happen anymore with this patch.

**Comment by RJVB** [comment]
Inline copy for easier glancing over:

```
diff --git a/src/qmplay2/VideoFilters.cpp b/src/qmplay2/VideoFilters.cpp
index 411299f48bc344d17d74da960a16e29445c1b786..fe299839bbc447c3ca5724a7df33087e3ddad5fc 100644
--- a/src/qmplay2/VideoFilters.cpp
+++ b/src/qmplay2/VideoFilters.cpp
@@ -143,7 +143,7 @@ private:
 class VideoFiltersThr final : public QThread
 {
 public:
-	VideoFiltersThr(VideoFilters &videoFilters) :
+	VideoFiltersThr(VideoFilters *videoFilters) :
 		videoFilters(videoFilters)
 	{
 		setObjectName("VideoFiltersThr");
@@ -181,7 +181,7 @@ public:
 		bufferMutex.lock();
 		while (filtering && !br)
 		{
-			if (!waitForAllFrames && !videoFilters.outputQueue.isEmpty())
+			if (!waitForAllFrames && !videoFilters->outputQueue.isEmpty())
 				break;
 			cond.wait(&bufferMutex);
 		}
@@ -209,7 +209,7 @@ private:
 			bool pending = false;
 			do
 			{
-				for (VideoFilter *vFilter : videoFilters.filters)
+				for (VideoFilter *vFilter : videoFilters->filters)
 				{
 					pending |= vFilter->filter(queue);
 					if (queue.isEmpty())
@@ -223,8 +223,8 @@ private:
 					QMutexLocker locker(&bufferMutex);
 					if (!queue.isEmpty())
 					{
-						videoFilters.outputQueue.append(queue);
-						videoFilters.outputNotEmpty = true;
+						videoFilters->outputQueue.append(queue);
+						videoFilters->outputNotEmpty = true;
 						queue.clear();
 					}
 					if (!pending)
@@ -242,7 +242,7 @@ private:
 		}
 	}
 
-	VideoFilters &videoFilters;
+	VideoFilters *videoFilters = NULL;
 
 	bool br = false, filtering = false;
 
@@ -269,25 +269,25 @@ void VideoFilters::init()
 }
 
 VideoFilters::VideoFilters() :
-	filtersThr(*(new VideoFiltersThr(*this))),
+	filtersThr(new VideoFiltersThr(this)),
 	outputNotEmpty(false)
 {}
 VideoFilters::~VideoFilters()
 {
 	clear();
-	delete &filtersThr;
+	delete filtersThr;
 }
 
 void VideoFilters::start()
 {
 	if (!filters.isEmpty())
-		filtersThr.start();
+		filtersThr->start();
 }
 void VideoFilters::clear()
 {
 	if (!filters.isEmpty())
 	{
-		filtersThr.stop();
+		filtersThr->stop();
 		for (VideoFilter *vFilter : filters)
 			delete vFilter;
 		filters.clear();
@@ -326,7 +326,7 @@ void VideoFilters::clearBuffers()
 {
 	if (!filters.isEmpty())
 	{
-		filtersThr.waitForFinished(true);
+		filtersThr->waitForFinished(true);
 		for (VideoFilter *vFilter : filters)
 			vFilter->clearBuffer();
 	}
@@ -337,7 +337,7 @@ void VideoFilters::removeLastFromInputBuffer()
 {
 	if (!filters.isEmpty())
 	{
-		filtersThr.waitForFinished(true);
+		filtersThr->waitForFinished(true);
 		for (int i = filters.count() - 1; i >= 0; --i)
 			if (filters[i]->removeLastFromInternalBuffer())
 				break;
@@ -348,7 +348,7 @@ void VideoFilters::addFrame(const VideoFrame &videoFrame, double ts)
 {
 	const VideoFilter::FrameBuffer frame(videoFrame, ts);
 	if (!filters.isEmpty())
-		filtersThr.filterFrame(frame);
+		filtersThr->filterFrame(frame);
 	else
 	{
 		outputQueue.enqueue(frame);
@@ -359,7 +359,7 @@ bool VideoFilters::getFrame(VideoFrame &videoFrame, TimeStamp &ts)
 {
 	bool locked, ret;
 	if ((locked = !filters.isEmpty()))
-		filtersThr.waitForFinished(false);
+		filtersThr->waitForFinished(false);
 	if ((ret = !outputQueue.isEmpty()))
 	{
 		videoFrame = outputQueue.at(0).frame;
@@ -368,14 +368,14 @@ bool VideoFilters::getFrame(VideoFrame &videoFrame, TimeStamp &ts)
 		outputNotEmpty = !outputQueue.isEmpty();
 	}
 	if (locked)
-		filtersThr.bufferMutex.unlock();
+		filtersThr->bufferMutex.unlock();
 	return ret;
 }
 
 bool VideoFilters::readyRead()
 {
-	filtersThr.waitForFinished(false);
+	filtersThr->waitForFinished(false);
 	const bool ret = outputNotEmpty;
-	filtersThr.bufferMutex.unlock();
+	filtersThr->bufferMutex.unlock();
 	return ret;
 }
diff --git a/src/qmplay2/headers/VideoFilters.hpp b/src/qmplay2/headers/VideoFilters.hpp
index 70af7adb3d81c2ef5641a252fa5963b406bcf1a5..db16f16819099d0f62b985d2f48dfb3e9e9dea04 100644
--- a/src/qmplay2/headers/VideoFilters.hpp
+++ b/src/qmplay2/headers/VideoFilters.hpp
@@ -62,6 +62,6 @@ private:
 
 	QQueue<VideoFilter::FrameBuffer> outputQueue;
 	QVector<VideoFilter *> filters;
-	VideoFiltersThr &filtersThr;
+	VideoFiltersThr *filtersThr;
 	bool outputNotEmpty = false;
 };
```

**Comment by zaps166** [comment]
Weird, I don't understand. Changing a reference class member into a pointer class member fixes a crash? When it crashes - on starting/stopping, on playback, or is it random behavior? I also used Clang 5.0.1 on Linux and QMPlay2 doesn't crash. For me reference and pointer is the same for the CPU (except that the reference address is initialized once and it must be initialized; initializing a reference with `nullptr` is a very bad idea). How to reproduce it? Do you use VideoToolBox video decoder, or any special compiler flags (LTO, etc...?). Could you give me more information? It is interesting, I want to check e.g. the assembly code after crash (`disassemble` in lldb should give something + show CPU registers).

---

```c++
#include <string>

class Asd
{
public:
    __attribute__((noinline)) Asd()
        : m_str(*(new std::string("asd")))
    {

    }
    __attribute__((noinline)) ~Asd()
    {
        delete &m_str;
    }

private:
    std::string &m_str;
};

class Zxc
{
public:
    __attribute__((noinline)) Zxc()
        : m_str(new std::string("asd"))
    {

    }
    __attribute__((noinline)) ~Zxc()
    {
        delete m_str;
    }

private:
    std::string *m_str = nullptr;
};

int main()
{
    Asd asd;
    Zxc zxc;
    return 0;
}
```
Classes `Asd` and `Zxc` gives me the same assembly code (you can check it here: https://godbolt.org/, use e.g. `-O3 -std=c++11` flags).

**Comment by RJVB** [comment]
>Weird, I don't understand. 

I also don't understand, because as you say, a reference variable is just a pointer variable made to look like it's not.  It also means I don't really see the point in what you are doing here, but that's another issue ;) Anyway, exactly the reason why I removed the extra steps (indirections) so the code is easier to follow during debugging. 

> How to reproduce it? 

I don't know. I forgot to mention: this happens apparently at random when I start a video (always the 1st time). It happens just a bit too often to write off to some freak event related to some playback initialisation glitch. Multimedia apps using cross-platform libraries sometimes crash for no apparent reason; even Qt apps are vulnerable to seemingly inexplicable crashes accessing member functions of class instances that were already deleted. Often that involves member variabes that are ObjC objects which outlived their parent C++ class.
That doesn't appear to be the case here though, but I haven't (yet) been able to see exactly when during the playback starting process the crash occurs. You may have a better idea, even without a full backtrace?

I build QMPlay2 with support for VideoToolBox, but it never worked for me AFAICR (not just in QMPlay). I will have to look into that again. I don't use LTO (strips debug info on Mac) but for the rest I use the same compiler flags as on Linux. Notably -Ofast; that should only affect math operations, no?

I have rebuilt QMPlay on Linux with clang 5.0.1 now, and even with -stdlib=libc++ so it uses at least partly the same C++ runtime. Sadly new & delete will probably still come from libstdc++ on Linux and libc++abi on Mac, but it seems unlikely that new() will return NULL because it isn't supposed to do that.

I'll keep you posted, and will try to post a backtrace (without the patch) ASAP.

godbolt.org doesn't seem to have clang 5.0.1 yet, but that aside: nice!

**Comment by zaps166** [comment]
@RJVB Do you have anything new about this bug?

**Comment by RJVB** [comment]
Nope, sorry. I have reversed my patch to see if I can reproduce the crash but haven't been using QMPlay2 a lot. Remains weird...

**Comment by zaps166** [comment]
Ok.

I can just switch it to pointer, but I',m so curious 😄

---

Btw. Why you use normal Clang 5.0.1 on macOS instead of weird Apple Clang version 8.x, 9.x from command line tools and/or Xcode ? Is it better on macOS?

**Comment by RJVB** [comment]
> I can just switch it to pointer, but I',m so curious 😄

I'm curious too. But also why you use this approach.

> Btw. Why you use normal Clang 5.0.1 on macOS instead of weird Apple Clang version 8.x, 9.x from command line tools and/or Xcode ? Is it better on macOS?

Apple's Clang is coupled to Xcode which is coupled to your OS version. So if you want a newer compiler you need to upgrade the whole OS (which tends to break and/or change a lot of things). And even then you rarely get the latest compiler version, and you don't get all the LLVM libraries and headers. So yeah, using a compiler from an independent source has advantages.

**Comment by zaps166** [comment]
> But also why you use this approach.

I wanted to hide `VideoFiltersThr` class, because it is a private class, so no need to set symbols to visible (I wanted to hide symbols). Previously I used stack object of this class, so changing it into a reference was faster (no need to change every usage of the variable from `.` to `->`). That's all 😄

---

> but it seems unlikely that new() will return NULL

`new` can return `nullptr` only if out of memory and it is called with `nothrow`:
```c++ 
new (std::nothrow) Something
```
http://www.cplusplus.com/reference/new/nothrow/

Otherwise it is probably a compiler or standard library bug or somebody overrides a `new` operator globally.

---

Maybe compiler doesn't assign the memory address value or doesn't even call a `new` operator? But you wrote that it doesn't happen always. Check if deinterlacing filters are still enabled in settings (filter will be created, but not used if video is not interlaced by default). Also add some `qDebug() << Q_FUNC_INFO << filters;` into methods `VideoFilters::clear()` and `VideoFilters::start()` to check if it can be called. Next time, you can also check `qDebug() << &filtersThr;` in these methods. You can send me console output :)

**Comment by RJVB** [comment]
I finally got a crash, so here's a post-mortem backtrace (courtesy of DrKonqi :))

The crashed thread is #16

I do notice that the LaunchServices picked up the executable in the build directory. That's Mac OS for you...
Of course the executable *should* be the same but, well, another thing out of my control here.


```
Application: QMPlay2 (QMPlay2), signal: Segmentation fault: 11
(lldb) process attach --pid 18964
Process 18964 stopped
Executable module set to "/opt/local/var/macports/build/_opt_local_site-ports_multimedia_qmplay2/qmplay2-devel/work/build/src/gui/QMPlay2.app/Contents/MacOS/QMPlay2".
Architecture set to: x86_64-apple-macosx.
(lldb) set term-width 200
(lldb) thread info
thread #1: tid = 0x3221079, 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP

(lldb) bt all
* thread #1: tid = 0x3221079, 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff8b8ced18 libsystem_kernel.dylib`mach_msg + 64
    frame #2: 0x00007fff96286f15 CoreFoundation`__CFRunLoopServiceMachPort + 181
    frame #3: 0x00007fff96286539 CoreFoundation`__CFRunLoopRun + 1161
    frame #4: 0x00007fff96285e75 CoreFoundation`CFRunLoopRunSpecific + 309
    frame #5: 0x00007fff8e7c6a0d HIToolbox`RunCurrentEventLoopInMode + 226
    frame #6: 0x00007fff8e7c67b7 HIToolbox`ReceiveNextEventCommon + 479
    frame #7: 0x00007fff8e7c65bc HIToolbox`_BlockUntilNextEventMatchingListInModeWithFilter + 65
    frame #8: 0x00007fff8abee24e AppKit`_DPSNextEvent + 1434
    frame #9: 0x00007fff8abed89b AppKit`-[NSApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 122
    frame #10: 0x00007fff8abe199c AppKit`-[NSApplication run] + 553
    frame #11: 0x00000001112145f1 libqaltcocoa.dylib`QCocoaEventDispatcher::processEvents(this=0x000000010d4cb4b0, flags=<unavailable>) + 3297 at qcocoaeventdispatcher.mm:431
    frame #12: 0x000000010776b981 QtCore`QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) [inlined] QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 401 at qeventloop.cpp:134
    frame #13: 0x000000010776b964 QtCore`QEventLoop::exec(this=0x00007fff599b1898, flags=<unavailable>) + 372 at qeventloop.cpp:212
    frame #14: 0x0000000107770412 QtCore`QCoreApplication::exec() + 402 at qcoreapplication.cpp:1298
    frame #15: 0x0000000106255b65 QMPlay2`main(argc=1, argv=<unavailable>) + 8453 at Main.cpp:724
    frame #16: 0x00007fff8c43d5fd libdyld.dylib`start + 1

  thread #2: tid = 0x32210a7, 0x00007fff8b8d4662 libsystem_kernel.dylib`kevent64 + 10, queue = 'com.apple.libdispatch-manager'
    frame #0: 0x00007fff8b8d4662 libsystem_kernel.dylib`kevent64 + 10
    frame #1: 0x00007fff8fc23421 libdispatch.dylib`_dispatch_mgr_invoke + 239
    frame #2: 0x00007fff8fc23136 libdispatch.dylib`_dispatch_mgr_thread + 52

  thread #3: tid = 0x32210a8, 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #0: 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff899d9f08 libsystem_pthread.dylib`_pthread_wqthread + 330
    frame #2: 0x00007fff899dcfb9 libsystem_pthread.dylib`start_wqthread + 13

  thread #4: tid = 0x32210a9, 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #0: 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff899d9f08 libsystem_pthread.dylib`_pthread_wqthread + 330
    frame #2: 0x00007fff899dcfb9 libsystem_pthread.dylib`start_wqthread + 13

  thread #5: tid = 0x32210b8, 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #0: 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff899d9f08 libsystem_pthread.dylib`_pthread_wqthread + 330
    frame #2: 0x00007fff899dcfb9 libsystem_pthread.dylib`start_wqthread + 13

  thread #6: tid = 0x32210be, 0x00007fff8b8d494a libsystem_kernel.dylib`poll + 10, name = 'QDBusConnectionManager'
    frame #0: 0x00007fff8b8d494a libsystem_kernel.dylib`poll + 10
    frame #1: 0x00000001077c4e60 QtCore`qt_safe_poll(pollfd*, unsigned int, timespec const*) [inlined] qt_ppoll(fds=0x0000000116d11438, nfds=2) + 592 at qcore_unix.cpp:83
    frame #2: 0x00000001077c4e55 QtCore`qt_safe_poll(fds=0x0000000116d11438, nfds=2, timeout_ts=<unavailable>) + 581 at qcore_unix.cpp:102
    frame #3: 0x00000001077c6350 QtCore`QEventDispatcherUNIX::processEvents(this=<unavailable>, flags=<unavailable>) + 880 at qeventdispatcher_unix.cpp:500
    frame #4: 0x000000010776b981 QtCore`QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) [inlined] QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 401 at qeventloop.cpp:134
    frame #5: 0x000000010776b964 QtCore`QEventLoop::exec(this=0x0000000116af6e28, flags=<unavailable>) + 372 at qeventloop.cpp:212
    frame #6: 0x000000010759e6e5 QtCore`QThread::exec(this=<unavailable>) + 117 at qthread.cpp:515
    frame #7: 0x0000000114b2abf3 QtDBus`QDBusConnectionManager::run(this=0x0000000115f51a40) + 19 at qdbusconnection.cpp:178
    frame #8: 0x00000001075a24e0 QtCore`QThreadPrivate::start(arg=0x0000000115f51a40) + 368 at qthread_unix.cpp:368
    frame #9: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #10: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #11: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #7: tid = 0x32210bf, 0x00007fff8b8d39aa libsystem_kernel.dylib`__select + 10, name = 'com.apple.CFSocket.private'
    frame #0: 0x00007fff8b8d39aa libsystem_kernel.dylib`__select + 10
    frame #1: 0x00007fff962d2a03 CoreFoundation`__CFSocketManager + 867
    frame #2: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #3: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #4: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #8: tid = 0x3221113, 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #0: 0x00007fff8b8d3e6a libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff899d9f08 libsystem_pthread.dylib`_pthread_wqthread + 330
    frame #2: 0x00007fff899dcfb9 libsystem_pthread.dylib`start_wqthread + 13

  thread #9: tid = 0x322111e, 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10
    frame #0: 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff8b8ced18 libsystem_kernel.dylib`mach_msg + 64
    frame #2: 0x00007fff96286f15 CoreFoundation`__CFRunLoopServiceMachPort + 181
    frame #3: 0x00007fff96286539 CoreFoundation`__CFRunLoopRun + 1161
    frame #4: 0x00007fff96285e75 CoreFoundation`CFRunLoopRunSpecific + 309
    frame #5: 0x00007fff8ad8e05e AppKit`_NSEventThread + 144
    frame #6: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #7: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #8: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #10: tid = 0x3221124, 0x00007fff8b8d3a3a libsystem_kernel.dylib`__semwait_signal + 10, name = 'DemuxerThr'
    frame #0: 0x00007fff8b8d3a3a libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff959d5dcc libsystem_c.dylib`nanosleep + 200
    frame #2: 0x00000001077c47db QtCore`qt_nanosleep(amount=(tv_sec = 0, tv_nsec = 15000000)) + 43 at qelapsedtimer_mac.cpp:107
    frame #3: 0x000000010628ef84 QMPlay2`DemuxerThr::run(this=0xbff000000000004b) + 7908 at DemuxerThr.cpp:557
    frame #4: 0x00000001075a24e0 QtCore`QThreadPrivate::start(arg=0x000000011fe62fc0) + 368 at qthread_unix.cpp:368
    frame #5: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #6: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #7: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #11: tid = 0x3221128, 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #0: 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff899dac3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x000000010849015d libavcodec.57.dylib`___lldb_unnamed_function3595$$libavcodec.57.dylib + 141
    frame #3: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #4: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #5: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #12: tid = 0x3221129, 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #0: 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff899dac3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x000000010849015d libavcodec.57.dylib`___lldb_unnamed_function3595$$libavcodec.57.dylib + 141
    frame #3: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #4: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #5: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #13: tid = 0x322112a, 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #0: 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff899dac3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x000000010849015d libavcodec.57.dylib`___lldb_unnamed_function3595$$libavcodec.57.dylib + 141
    frame #3: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #4: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #5: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #14: tid = 0x322112b, 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #0: 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff899dac3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x000000010849015d libavcodec.57.dylib`___lldb_unnamed_function3595$$libavcodec.57.dylib + 141
    frame #3: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #4: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #5: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #15: tid = 0x322112c, 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #0: 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff899dac3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x000000010849015d libavcodec.57.dylib`___lldb_unnamed_function3595$$libavcodec.57.dylib + 141
    frame #3: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #4: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #5: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #16: tid = 0x322112d, 0x00007fff8b8d3e22 libsystem_kernel.dylib`__wait4 + 10, name = 'VideoThr'
    frame #0: 0x00007fff8b8d3e22 libsystem_kernel.dylib`__wait4 + 10
    frame #1: 0x000000011646ad1e libKF5Crash.5.dylib`KCrash::startProcess(int, char const**, bool) [inlined] startProcessInternal(argc=<unavailable>, directly=<unavailable>) + 125 at kcrash.cpp:649
    frame #2: 0x000000011646aca1 libKF5Crash.5.dylib`KCrash::startProcess(argc=<unavailable>, argv=<unavailable>, waitAndExit=<unavailable>) + 17 at kcrash.cpp:631
    frame #3: 0x000000011646aae5 libKF5Crash.5.dylib`KCrash::defaultCrashHandler(sig=11) + 1061 at kcrash.cpp:528
    frame #4: 0x00007fff8e69b5aa libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x000000010759a85f QtCore`QMutex::lock() [inlined] std::__1::__atomic_base<QMutexData*, false>::compare_exchange_strong(__s=<unavailable>, __f=<unavailable>) + 1 at atomic:606
    frame #6: 0x000000010759a85e QtCore`QMutex::lock() [inlined] bool QAtomicOps<QMutexData*>::testAndSetAcquire<QMutexData*>(std::__1::atomic<QMutexData*>&, QMutexData*, QMutexData*, QMutexData**) at qatomic_cxx11.h:290
    frame #7: 0x000000010759a85e QtCore`QMutex::lock() [inlined] QBasicAtomicPointer<QMutexData>::testAndSetAcquire(QMutexData*, QMutexData*, QMutexData*&) at qbasicatomic.h:262
    frame #8: 0x000000010759a85e QtCore`QMutex::lock() [inlined] QBasicMutex::fastTryLock(QMutexData*&) at qmutex.h:101
    frame #9: 0x000000010759a85e QtCore`QMutex::lock(this=0x0000000000000010) + 14 at qmutex.cpp:224
    frame #10: 0x0000000106603e6e libqmplay2.dylib`VideoFilters::readyRead() [inlined] VideoFiltersThr::waitForFinished(this=0x0000000000000000, waitForAllFrames=<unavailable>) + 12 at VideoFilters.cpp:181
    frame #11: 0x0000000106603e62 libqmplay2.dylib`VideoFilters::readyRead(this=0x000000011ebc0998) + 18 at VideoFilters.cpp:377
    frame #12: 0x00000001062957c1 QMPlay2`VideoThr::run(this=0x000000011ebc0920) + 673 at VideoThr.cpp:274
    frame #13: 0x00000001075a24e0 QtCore`QThreadPrivate::start(arg=0x000000011ebc0920) + 368 at qthread_unix.cpp:368
    frame #14: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #15: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #16: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #17: tid = 0x322112e, 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'VideoFiltersThr'
    frame #0: 0x00007fff8b8d3716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff899dac3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x00000001075a3b6b QtCore`QWaitConditionPrivate::wait(this=0x0000000120b75900, time=<unavailable>) + 75 at qwaitcondition_unix.cpp:143
    frame #3: 0x00000001075a3a12 QtCore`QWaitCondition::wait(this=0x0000000120953688, mutex=0x0000000120953690, time=18446744073709551615) + 162 at qwaitcondition_unix.cpp:215
    frame #4: 0x0000000106604164 libqmplay2.dylib`VideoFiltersThr::run(this=0x0000000120953660) + 148 at VideoFilters.cpp:201
    frame #5: 0x00000001075a24e0 QtCore`QThreadPrivate::start(arg=0x0000000120953660) + 368 at qthread_unix.cpp:368
    frame #6: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #7: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #8: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #18: tid = 0x3221130, 0x00007fff8b8d3a3a libsystem_kernel.dylib`__semwait_signal + 10, name = 'AudioThr'
    frame #0: 0x00007fff8b8d3a3a libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff959d5dcc libsystem_c.dylib`nanosleep + 200
    frame #2: 0x000000011783f5fe libportaudio.2.dylib`Pa_Sleep(msec=<unavailable>) + 78 at pa_unix_util.c:116
    frame #3: 0x0000000117845336 libportaudio.2.dylib`WriteStream(stream=0x0000000120932400, buffer=0x0000000124be3560, framesRequested=788) + 86 at pa_mac_core_blocking.c:538
    frame #4: 0x0000000117823e85 libPortAudio.dylib`PortAudioWriter::write(QByteArray const&) [inlined] PortAudioWriter::writeStream(this=0x0000000120bd4600) + 33 at PortAudioWriter.cpp:285
    frame #5: 0x0000000117823e64 libPortAudio.dylib`PortAudioWriter::write(this=0x0000000120bd4600, arr=0x00000001243c7d48) + 148 at PortAudioWriter.cpp:181
    frame #6: 0x000000010629ab93 QMPlay2`AudioThr::run(this=<unavailable>) + 4915 at AudioThr.cpp:392
    frame #7: 0x00000001075a24e0 QtCore`QThreadPrivate::start(arg=0x0000000120945980) + 368 at qthread_unix.cpp:368
    frame #8: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #9: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #10: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13

  thread #19: tid = 0x3221149, 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10, name = 'com.apple.audio.IOThread.client'
    frame #0: 0x00007fff8b8cfa1a libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff8b8ced18 libsystem_kernel.dylib`mach_msg + 64
    frame #2: 0x00007fff942457fc CoreAudio`HALB_MachPort::SendMessageWithReply(unsigned int, unsigned int, unsigned int, unsigned int, mach_msg_header_t*, bool, unsigned int) + 98
    frame #3: 0x00007fff9424578a CoreAudio`HALB_MachPort::SendSimpleMessageWithSimpleReply(unsigned int, unsigned int, int, int&, bool, unsigned int) + 42
    frame #4: 0x00007fff94243efc CoreAudio`HALC_ProxyIOContext::IOWorkLoop() + 1140
    frame #5: 0x00007fff942439dd CoreAudio`HALC_ProxyIOContext::IOThreadEntry(void*) + 97
    frame #6: 0x00007fff9424389d CoreAudio`HALB_IOThread::Entry(void*) + 75
    frame #7: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #8: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #9: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13
(lldb) detach
Process 18964 detached
```

**Comment by zaps166** [comment]
Ok, thanks. It should crash earlier in `VideoFilters::start()` if any filter is available. How "Video filters" tab looks on your macOS?
![screenshot_20180126_001130](https://user-images.githubusercontent.com/6239090/35417389-7688deae-022d-11e8-94e4-8f1fe9149b73.png)
If `Remove interlacing` checkbox is checked then it is a very weird crash, probably use-after-free (why?) or buffer overflow somewhere or even a compiler bug...

`VideoFilters::start()` **must** crash when `&filtersThr` is `nullptr` and `filters` is not empty.

---

A long time before changes in `VideoFilters` class I've noticed a crash on macOS when starting video playback (maybe 2 o 3 times). Maybe it was similar crash (I don't remember).

---

Any chance for `qDebug()` as described in https://github.com/zaps166/QMPlay2/issues/137#issuecomment-360270213 ? :-) Also for `VideoFilters::readyRead()`.

**Comment by zaps166** [comment]
Also disassembly of e.g. `VideoFilters::readyRead()` will be fine (disassembly of method which calls `nullptr`) - in case of previous backtrace it is thread 16, frame 11, iirc `select frame 11` and `disassemble`.

**Comment by RJVB** [comment]
> How "Video filters" tab looks on your macOS?
> ![screenshot_20180126_001130](https://user-images.githubusercontent.com/6239090/35417389-7688deae-022d-11e8-94e4-8f1fe9149b73.png)

Like that but without the VDPAU and VA-API options of course (and in my case parity is set to bottom field first, I can't remember having done that myself).

> If `Remove interlacing` checkbox is checked then it is a very weird crash, probably use-after-free (why?) or buffer overflow somewhere or even a compiler bug...

I've also thought of a use-after-free bug, or else something that overwrites the stack (writing out of bounds in a C string for instance). Or maybe there's a function somewhere that lacks a return value; I recently saw an example of the kind of crash that can give.

> A long time before changes in `VideoFilters` class I've noticed a crash on macOS when starting video playback (maybe 2 o 3 times). Maybe it was similar crash (I don't remember).

This also happens when starting playback.

I'll add the debug output, and cross our fingers that this will not make the bug jump to a completely different place (you know, a Heisenbug ;)) 

Meanwhile, I managed to get another crash with the current build:

```
* thread #14: tid = 0x32cec9d, 0x00007fff8b8d3e22 libsystem_kernel.dylib`__wait4 + 10, name = 'VideoThr'
  * frame #0: 0x00007fff8b8d3e22 libsystem_kernel.dylib`__wait4 + 10
    frame #1: 0x000000011f362d1e libKF5Crash.5.dylib`KCrash::startProcess(int, char const**, bool) [inlined] startProcessInternal(argc=<unavailable>, directly=<unavailable>) + 125 at kcrash.cpp:649
    frame #2: 0x000000011f362ca1 libKF5Crash.5.dylib`KCrash::startProcess(argc=<unavailable>, argv=<unavailable>, waitAndExit=<unavailable>) + 17 at kcrash.cpp:631
    frame #3: 0x000000011f362ae5 libKF5Crash.5.dylib`KCrash::defaultCrashHandler(sig=11) + 1061 at kcrash.cpp:528
    frame #4: 0x00007fff8e69b5aa libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x00000001112a485f QtCore`QMutex::lock() [inlined] std::__1::__atomic_base<QMutexData*, false>::compare_exchange_strong(__s=<unavailable>, __f=<unavailable>) + 1 at atomic:606
    frame #6: 0x00000001112a485e QtCore`QMutex::lock() [inlined] bool QAtomicOps<QMutexData*>::testAndSetAcquire<QMutexData*>(std::__1::atomic<QMutexData*>&, QMutexData*, QMutexData*, QMutexData**) at qatomic_cxx11.h:290
    frame #7: 0x00000001112a485e QtCore`QMutex::lock() [inlined] QBasicAtomicPointer<QMutexData>::testAndSetAcquire(QMutexData*, QMutexData*, QMutexData*&) at qbasicatomic.h:262
    frame #8: 0x00000001112a485e QtCore`QMutex::lock() [inlined] QBasicMutex::fastTryLock(QMutexData*&) at qmutex.h:101
    frame #9: 0x00000001112a485e QtCore`QMutex::lock(this=0x0000000000000010) + 14 at qmutex.cpp:224
    frame #10: 0x000000011030ae6e libqmplay2.dylib`VideoFilters::readyRead() [inlined] VideoFiltersThr::waitForFinished(this=0x0000000000000000, waitForAllFrames=<unavailable>) + 12 at VideoFilters.cpp:181
    frame #11: 0x000000011030ae62 libqmplay2.dylib`VideoFilters::readyRead(this=0x000000012888afd8) + 18 at VideoFilters.cpp:377
    frame #12: 0x000000010ffa37c1 QMPlay2`VideoThr::run(this=0x000000012888af60) + 673 at VideoThr.cpp:274
    frame #13: 0x00000001112ac4e0 QtCore`QThreadPrivate::start(arg=0x000000012888af60) + 368 at qthread_unix.cpp:368
    frame #14: 0x00007fff899d8899 libsystem_pthread.dylib`_pthread_body + 138
    frame #15: 0x00007fff899d872a libsystem_pthread.dylib`_pthread_start + 137
    frame #16: 0x00007fff899dcfc9 libsystem_pthread.dylib`thread_start + 13
```

```
(lldb) f 10
frame #10: 0x000000011030ae6e libqmplay2.dylib`VideoFilters::readyRead() [inlined] VideoFiltersThr::waitForFinished(this=0x0000000000000000, waitForAllFrames=<unavailable>) + 12 at VideoFilters.cpp:181
   178  
   179          void waitForFinished(bool waitForAllFrames)
   180          {
-> 181                  bufferMutex.lock();
   182                  while (filtering && !br)
   183                  {
   184                          if (!waitForAllFrames && !videoFilters->outputQueue.isEmpty())
(lldb) disassemble
libqmplay2.dylib`VideoFilters::readyRead() at VideoFilters.cpp:376:
   0x11030ae50:  pushq  %rbp
   0x11030ae51:  movq   %rsp, %rbp
   0x11030ae54:  pushq  %r15
   0x11030ae56:  pushq  %r14
   0x11030ae58:  pushq  %r12
   0x11030ae5a:  pushq  %rbx
   0x11030ae5b:  movq   %rdi, %r14
   0x11030ae5e:  movq   0x10(%r14), %rbx
   0x11030ae62:  leaq   0x10(%rbx), %r15
   0x11030ae66:  movq   %r15, %rdi
   0x11030ae69:  callq  0x110321018               ; symbol stub for: QMutex::lock()
-> 0x11030ae6e:  cmpb   $0x0, 0x21(%rbx)
   0x11030ae72:  je     0x11030aead               ; VideoFilters::readyRead() + 93 at VideoFilters.cpp:378
   0x11030ae74:  leaq   0x28(%rbx), %r12
   0x11030ae78:  nopl   (%rax,%rax)
   0x11030ae80:  cmpb   $0x0, 0x20(%rbx)
   0x11030ae84:  jne    0x11030aead               ; VideoFilters::readyRead() + 93 at VideoFilters.cpp:378
   0x11030ae86:  movq   0x18(%rbx), %rax
   0x11030ae8a:  movq   (%rax), %rax
   0x11030ae8d:  movl   0xc(%rax), %ecx
   0x11030ae90:  cmpl   0x8(%rax), %ecx
   0x11030ae93:  jne    0x11030aead               ; VideoFilters::readyRead() + 93 at VideoFilters.cpp:378
   0x11030ae95:  movq   $-0x1, %rdx
   0x11030ae9c:  movq   %r12, %rdi
   0x11030ae9f:  movq   %r15, %rsi
   0x11030aea2:  callq  0x110320f64               ; symbol stub for: QWaitCondition::wait(QMutex*, unsigned long)
   0x11030aea7:  cmpb   $0x0, 0x21(%rbx)
   0x11030aeab:  jne    0x11030ae80               ; VideoFilters::readyRead() + 48 [inlined] VideoFiltersThr::waitForFinished(bool) + 30 at VideoFilters.cpp:377
VideoFilters::readyRead() + 18 at VideoFilters.cpp:377
   0x11030aead:  movb   0x18(%r14), %bl
   0x11030aeb1:  movq   0x10(%r14), %rdi
   0x11030aeb5:  addq   $0x10, %rdi
   0x11030aeb9:  callq  0x11032101e               ; symbol stub for: QMutex::unlock()
   0x11030aebe:  movl   %ebx, %eax
   0x11030aec0:  popq   %rbx
   0x11030aec1:  popq   %r12
   0x11030aec3:  popq   %r14
   0x11030aec5:  popq   %r15
   0x11030aec7:  popq   %rbp
   0x11030aec8:  retq   

libqmplay2.dylib`VideoFilters::readyRead() + 121:
   0x11030aec9:  nopl   (%rax)
```

```
(lldb) f 11
frame #11: 0x000000011030ae62 libqmplay2.dylib`VideoFilters::readyRead(this=0x000000012888afd8) + 18 at VideoFilters.cpp:377
   374  
   375  bool VideoFilters::readyRead()
   376  {
-> 377          filtersThr->waitForFinished(false);
   378          const bool ret = outputNotEmpty;
   379          filtersThr->bufferMutex.unlock();
   380          return ret;
(lldb) disassemble
libqmplay2.dylib`VideoFilters::readyRead() at VideoFilters.cpp:376:
   0x11030ae50:  pushq  %rbp
   0x11030ae51:  movq   %rsp, %rbp
   0x11030ae54:  pushq  %r15
   0x11030ae56:  pushq  %r14
   0x11030ae58:  pushq  %r12
   0x11030ae5a:  pushq  %rbx
   0x11030ae5b:  movq   %rdi, %r14
   0x11030ae5e:  movq   0x10(%r14), %rbx
-> 0x11030ae62:  leaq   0x10(%rbx), %r15
```

**Comment by zaps166** [comment]
Thanks!

>  I'll add the debug output, and cross our fingers that this will not make the bug jump to a completely different place (you know, a Heisenbug ;))

Ok, I'll wait for this :) I know about this, but I didn't know the name (Heisenbug) :)

**Comment by RJVB** [comment]
Good.

Do you have a brilliant and simple idea how to get that qDebug() output to show up when I launch QMPlay2 via the Finder? In older Mac OS versions everything such apps printed on stdout and stderr went into the system.log but that's no longer the case.

Maybe you can whip up a consol/logviewer feature like VLC and Audacious have? ;)

R.

**Comment by zaps166** [comment]
Change message handler in `Main.cpp` file (redirect to file):
```c++
static void messageHandler(QtMsgType type, const QMessageLogContext &context, const QString &message)
{
	static FILE *f = fopen((QDir::homePath() + "/qmplay2log.txt").toUtf8().constData(), "w");
	fprintf(f, "%s\n", qFormatLogMessage(type, context, message).toLocal8Bit().constData());
	fflush(f);
}
```

**Comment by RJVB** [comment]
> Change message handler in `Main.cpp` file (redirect to file):

Or do this ;)

@@ -459,8 +462,19 @@ static LRESULT CALLBACK MMKeysHookProc(int code, WPARAM wparam, LPARAM lparam)
 #ifndef Q_OS_ANDROID
 static void messageHandler(QtMsgType type, const QMessageLogContext &context, const QString &message)
 {
-	fprintf(stderr, "%s\n", qFormatLogMessage(type, context, message).toLocal8Bit().constData());
+	const char *messageText = qFormatLogMessage(type, context, message).toLocal8Bit().constData();
+	fprintf(stderr, "%s\n", messageText);
 	fflush(stderr);
+#if defined Q_OS_MACOS
+	static const QMap<QtMsgType,int> type2Level = {
+		{QtDebugMsg, LOG_DEBUG},
+		{QtWarningMsg, LOG_WARNING},
+		{QtCriticalMsg, LOG_CRIT},
+		{QtFatalMsg, LOG_CRIT},
+		{QtInfoMsg, LOG_INFO}
+	};
+	syslog(LOG_NDELAY|LOG_PID|type2Level[type], "%s", messageText);
+#endif
 }
 #endif
 


(btw, Q_OS_MACOS is the preferred platform ID macro; Q_OS_MAC is provided for backwards compatibility).

R.

**Comment by zaps166** [comment]
I don't know why it crashes.
@RJVB Do you have some logs? :) If logs has nothing special and if pointer fixes a crash for you, just create a PR with https://github.com/zaps166/QMPlay2/issues/137#issuecomment-357425564

**Comment by RJVB** [comment]
> I don't know why it crashes.

You did well to ask. I hadn't been able to get it to crash for a while now, so I tried again.

The backtrace looks as usual, again from the copy in my build directory:

```
  thread #12: tid = 0x30d26, 0x00007fff8ec89e22 libsystem_kernel.dylib`__wait4 + 10, name = 'VideoThr'
    frame #0: 0x00007fff8ec89e22 libsystem_kernel.dylib`__wait4 + 10
    frame #1: 0x000000011455bd1e libKF5Crash.5.dylib`KCrash::startProcess(int, char const**, bool) [inlined] startProcessInternal(argc=<unavailable>, directly=<unavailable>) + 125 at kcrash.cpp:649
    frame #2: 0x000000011455bca1 libKF5Crash.5.dylib`KCrash::startProcess(argc=<unavailable>, argv=<unavailable>, waitAndExit=<unavailable>) + 17 at kcrash.cpp:631
    frame #3: 0x000000011455bae5 libKF5Crash.5.dylib`KCrash::defaultCrashHandler(sig=11) + 1061 at kcrash.cpp:528
    frame #4: 0x00007fff91a515aa libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x00000001056e39cf QtCore`QMutex::lock() [inlined] std::__1::__atomic_base<QMutexData*, false>::compare_exchange_strong(__s=<unavailable>, __f=<unavailable>) + 1 at atomic:606
    frame #6: 0x00000001056e39ce QtCore`QMutex::lock() [inlined] bool QAtomicOps<QMutexData*>::testAndSetAcquire<QMutexData*>(std::__1::atomic<QMutexData*>&, QMutexData*, QMutexData*, QMutexData**) at qatomic_cxx11.h:290
    frame #7: 0x00000001056e39ce QtCore`QMutex::lock() [inlined] QBasicAtomicPointer<QMutexData>::testAndSetAcquire(QMutexData*, QMutexData*, QMutexData*&) at qbasicatomic.h:262
    frame #8: 0x00000001056e39ce QtCore`QMutex::lock() [inlined] QBasicMutex::fastTryLock(QMutexData*&) at qmutex.h:101
    frame #9: 0x00000001056e39ce QtCore`QMutex::lock(this=0x0000000000000010) + 14 at qmutex.cpp:224
    frame #10: 0x000000010474db8e libqmplay2.dylib`VideoFilters::readyRead() [inlined] VideoFiltersThr::waitForFinished(this=0x0000000000000000, waitForAllFrames=<unavailable>) + 12 at VideoFilters.cpp:183
    frame #11: 0x000000010474db82 libqmplay2.dylib`VideoFilters::readyRead(this=0x000000011e9a0118) + 18 at VideoFilters.cpp:381
    frame #12: 0x00000001043e6791 QMPlay2`VideoThr::run(this=0x000000011e9a00a0) + 673 at VideoThr.cpp:274
    frame #13: 0x00000001056eb650 QtCore`QThreadPrivate::start(arg=0x000000011e9a00a0) + 368 at qthread_unix.cpp:368
    frame #14: 0x00007fff8cd8e899 libsystem_pthread.dylib`_pthread_body + 138
    frame #15: 0x00007fff8cd8e72a libsystem_pthread.dylib`_pthread_start + 137
    frame #16: 0x00007fff8cd92fc9 libsystem_pthread.dylib`thread_start + 13
```

but now I also have the debug output you requested:
Feb  4 16:48:12 Portia.local QMPlay2[30394]: void VideoFilters::clear() filters= QVector() &filtersThr= QThread(0x11e9a0280, name = "VideoFiltersThr")
Feb  4 16:48:12 Portia.local QMPlay2[30394]: void VideoFilters::start() filters= QVector(0x115c55d70) &filtersThr= QThread(0x11e9a0280, name = "VideoFiltersThr")

attaching a debugger and looking at the filtersThr object it appears to be fine itself, but both its QMutex members look suspicious:

  bufferMutex = {
    QBasicMutex = {
      d_ptr = {
        _q_value = {
          __a_ = 0x0000000000000000
        }
      }
    }
  }

However, I notice that my build directory also still contains old .so modules, e.g.

```
11006423 -rwxr-xr-x 1 bertin admin 72524 Jan 31 18:05 build/src/modules/VideoFilters/libVideoFilters.dylib*
832218 -rwxr-xr-x 1 bertin admin 94492 Apr  5  2017 build/src/modules/VideoFilters/libVideoFilters.so*
```

I'll see if the crash really goes away too when I remove those old .so files. If that's the case I owe you an apology, I should have realised earlier that
1) LaunchServices launches whatever copy it wants
2) the terminal error about mismatching module API versions doesn't mean QMPlay2 won't try to use them.

**Comment by zaps166** [comment]
According to backtrace: `QMutex` crashes, because it is member of `VideoFiltersThr` class and `filtersThr` points to `0x0`, a `bufferMutex` offset inside class is `0x10`, so `bufferMutex` has addres of `0x10` and it must crash (address is invalid). I don't have any code which directly sets `bufferMutex` to `nullptr`, so something must be wrong somewhere (where and why?).

> I'll see if the crash really goes away too when I remove those old .so files.

Why you have `*.so` file suffix on macOS :) ? I can see that QMPlay2 also loads file with `*.so` suffix, so maybe this can be a problem, but crash happens in `libqmplay2.dylib`. Maybe it loads both libraries into memory and once it jumps to correct library and sometimes it jumps to old library code and it crashes? 

Log output shows correct `VideoFiltersThr` address - does the log are soon before crash.

---

You can remove all compiled binaries and all other compilation stuff and try again.

**Comment by RJVB** [comment]
>According to backtrace: `QMutex` crashes, because it is member of `VideoFiltersThr` class and `filtersThr` points to `0x0`

Yes, but those are post-mortem values. I don't trust them completely in this case. Case in point: I can attach the debugger and print the value of filterThr one frame up from where it is supposedly a null pointer. That gives me the same value as what is shown in the log. To answer your question: I *think* the log output must be generated just before the crash happens. It comes from qDebug calls I added following your instructions. I'll see if I can add a few more.

>Why you have `*.so` file suffix on macOS :) ?

You get those when you build "MODULE libraries" with cmake. They're almost a year old, maybe you modified your build system to generate dylib files since?

>I can see that QMPlay2 also loads file with `*.so` suffix, so maybe this can be a problem, but crash happens in `libqmplay2.dylib`. Maybe it loads both libraries into memory and once it jumps to correct library and sometimes it jumps to old library code and it crashes? 

That could be an explanation. You know better than I how QMPlay2 handles files with the wrong API version. I'd hope the warning means that the file isn't loaded.

>You can remove all compiled binaries and all other compilation stuff and try again.

Yes, but that shouldn't be necessary. Of course one also shouldn't be running the executable from the build dir by accident, but that's Mac OS for you...

**Comment by zaps166** [comment]
> maybe you modified your build system to generate dylib files since?

Yes, maybe it was modified :) I don't remember, but probably yes.

> You know better than I how QMPlay2 handles files with the wrong API version.

`lib.setLoadHints(QLibrary::DeepBindHint);` maybe something is wrong with this (especially on macOS)? QMPlay2 loads a library and then checks for version. I didn't have problems on Linux (yet, not many tests, but it worked when mixing Qt4 and Qt5 libs - it shows error and it worked with another lib IIRC).

**Comment by zaps166** [comment]
>  Maybe you can whip up a consol/logviewer feature

Partially done, you can watch categorized and not categorized Qt logs (except debugging level `qDebug`, `qCDebug`) in QMPlay2 logger, it'll also write them to text file.

**Comment by RJVB** [comment]
Maybe this is indeed a case of memory (stack) corruption. I was testing the modified bitperfect code with QMPlay2 started from the terminal (= running the properly installed copy), starting various items in my playlist. I got this this when I double-clicked a video when an audio m4a was playing:

```
> QMPlay2
[...]
2018-02-06 21:31:13.178 QMPlay2[5536:d07] SetNominalSampleRate(48000) setting rate to 48000Hz
[qdraw @ 0x12bb79300] [IMGUTILS @ 0x119303558] Picture size 29044x31088 is invalid
KCrash: Application 'QMPlay2' crashing...
KCrash: Attempting to start /opt/local/libexec/kde5/drkonqi directly
void VideoFilters::clear() filters= QVector() &filtersThr= QThread(0x11c6a1fc0, name = "VideoFiltersThr")
void VideoFilters::start() filters= QVector() &filtersThr= QThread(0x11c6a1fc0, name = "VideoFiltersThr")
[ass] Shaper: FriBidi 0.19.6 (SIMPLE)
[ass] Using font provider coretext
2018-02-06 21:31:20.596 QMPlay2[5536:d07] SetNominalSampleRate(44100) setting rate to 44100Hz
2018-02-06 21:31:20.598 QMPlay2[5536:d07] AudioDevice USB PnP Sound Device (63) released
```

the 48000Hz line was from starting the m4a, the qdraw line under it appeared as soon as I launched the video, the `[ass]` lines and the 44100Hz lines were probably printed when DrKonqi released (SIGCONT signal) the application to get a backtrace (I've often noticed that playback would go on during that period, despite the crash). I'm not entirely certain when the filtersThr lines were printed, probably also after DrKonqi took control.

In this case there was no suspicious NULL this pointer, but the VideoFiltersThr mutex must have been corrupt all the same:

```
  thread #9: tid = 0x1198cd, 0x00007fff8ec89e22 libsystem_kernel.dylib`__wait4 + 10, name = 'VideoThr'
    frame #0: 0x00007fff8ec89e22 libsystem_kernel.dylib`__wait4 + 10
    frame #1: 0x00000001200fbd1e libKF5Crash.5.dylib`KCrash::startProcess(int, char const**, bool) [inlined] startProcessInternal(argc=<unavailable>, directly=<unavailable>) + 125 at kcrash.cpp:649
    frame #2: 0x00000001200fbca1 libKF5Crash.5.dylib`KCrash::startProcess(argc=<unavailable>, argv=<unavailable>, waitAndExit=<unavailable>) + 17 at kcrash.cpp:631
    frame #3: 0x00000001200fbae5 libKF5Crash.5.dylib`KCrash::defaultCrashHandler(sig=11) + 1061 at kcrash.cpp:528
    frame #4: 0x00007fff91a515aa libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x0000000110f4c9cf QtCore`QMutex::lock() [inlined] std::__1::__atomic_base<QMutexData*, false>::compare_exchange_strong(__s=<unavailable>, __f=<unavailable>) + 1 at atomic:606
    frame #6: 0x0000000110f4c9ce QtCore`QMutex::lock() [inlined] bool QAtomicOps<QMutexData*>::testAndSetAcquire<QMutexData*>(std::__1::atomic<QMutexData*>&, QMutexData*, QMutexData*, QMutexData**) at qatomic_cxx11.h:290
    frame #7: 0x0000000110f4c9ce QtCore`QMutex::lock() [inlined] QBasicAtomicPointer<QMutexData>::testAndSetAcquire(QMutexData*, QMutexData*, QMutexData*&) at qbasicatomic.h:262
    frame #8: 0x0000000110f4c9ce QtCore`QMutex::lock() [inlined] QBasicMutex::fastTryLock(QMutexData*&) at qmutex.h:101
    frame #9: 0x0000000110f4c9ce QtCore`QMutex::lock(this=0x0066002000720085) + 14 at qmutex.cpp:224
    frame #10: 0x000000010ffbb1e0 libqmplay2.dylib`VideoFilters::readyRead() [inlined] VideoFiltersThr::waitForFinished(this=0x0066002000720075, waitForAllFrames=<unavailable>) + 12 at VideoFilters.cpp:183
    frame #11: 0x000000010ffbb1d4 libqmplay2.dylib`VideoFilters::readyRead(this=0x000000012adbe6b8) + 356 at VideoFilters.cpp:387
    frame #12: 0x000000010fb3a541 QMPlay2`VideoThr::run(this=0x000000012adbe640) + 673 at VideoThr.cpp:279
    frame #13: 0x0000000110f54650 QtCore`QThreadPrivate::start(arg=0x000000012adbe640) + 368 at qthread_unix.cpp:368
    frame #14: 0x00007fff8cd8e899 libsystem_pthread.dylib`_pthread_body + 138
    frame #15: 0x00007fff8cd8e72a libsystem_pthread.dylib`_pthread_start + 137
    frame #16: 0x00007fff8cd92fc9 libsystem_pthread.dylib`thread_start + 13
```

lldb shows this in frame 11 of that thread:

```
(lldb) p filtersThr
(VideoFiltersThr) $1 = {
  QThread = {
    QObject = {
      d_ptr = {
        d = 0x0000000120a751e0
      }
    }
  }
  bufferMutex = {
    QBasicMutex = {
      d_ptr = {
        _q_value = {
          __a_ = 0x0000000000000000
        }
      }
    }
  }
  videoFilters = 0x000000012adbe6b8
  br = false
  filtering = false
  cond = {
    d = 0x000000012a403380
  }
  mutex = {
    QBasicMutex = {
      d_ptr = {
        _q_value = {
          __a_ = 0x0000000000000000
        }
      }
    }
  }
  frameToFilter = {
    frame = {
      size = (width = 0, height = 0, chromaShiftW = '\0', chromaShiftH = '\0')
      buffer = {
        [0] = {
          m_bufferRef = 0x0000000000000000
          m_size = 0
        }
        [1] = {
          m_bufferRef = 0x0000000000000000
          m_size = 0
        }
        [2] = {
          m_bufferRef = 0x0000000000000000
          m_size = 0
        }
      }
      linesize = ([0] = 0, [1] = 0, [2] = 0)
      interlaced = false
      tff = false
      surfaceId = 0
    }
    ts = 0
  }
}
(lldb) p *this
(VideoFilters) $2 = {
  outputQueue = <size=0>
  filters = <size=0>
  filtersThr = 0x000000011c6a1fc0
  outputNotEmpty = false
}
```

I haven't been able to reproduce the issue, as usual. Is there maybe a function somewhere that fails to return a value under certain circumstances?

**Comment by zaps166** [comment]
> Is there maybe a function somewhere that fails to return a value under certain circumstances?

Compiler should give me warning - I can't see any warnings.

---

Could you try to run QMPlay2 without any modifications especially w/o bit-perfect audio and w/o 3rdparty library modifications?

Once I had buffer overflow inside FFmpeg due to FFmpeg bug after FFmpeg update and it crashes somewhere in my `LibASS` class and `VideoThr` or `VideoFilters` (I don't remember exactly where) (workaround 3dc452f3de87f886fbd62a3aae94f2d770047a55). Buffer overflow can be difficult to catch.

**Comment by zaps166** [comment]
Maybe try `-fsanitize=undefined` and/or `-fsanitize=address`

**Comment by RJVB** [comment]
>Compiler should give me warning - I can't see any warnings.

Would it also warn about possible use of an unitialised QMutex? Come to think of it, I know Thiago modified something related to initialisation in the underlying QBasicMutex implementation, in Qt 5.10. I'm blocked on 5.9 though, for the moment.
I was supposed to be out today but snow and strikes are keeping me blocked at home so I'm going to look first into upgrading to Qt 5.9.4 (should be possible to upgrade only the main components and leave the real big ones like WebEngine for later).

>Could you try to run QMPlay2 without any modifications especially w/o bit-perfect audio and w/o 3rdparty library modifications?

The former yes, the latter probably not. I'll try the sanitise options, maybe this time I'll actually get them to work.

**Comment by RJVB** [comment]
> Maybe try `-fsanitize=undefined` and/or `-fsanitize=address`

Can't, sadly - not yet supported on my OS version.

I tried to build with GCC 7 coerced to use libc++ (where sanitisation does work), but that fails to link (linker crashes, apparently because of the presence of ObjC++ code).

Bummer.

**Comment by zaps166** [comment]
I use Qt 5.6.3 for QMPlay2 packages (macOS, Windows and Ubuntu(with modifications)), because it is stable enough.

---

> Can't, sadly - not yet supported on my OS version.

Which OS version do you have? Doesn't it depend on compiler and compiler tools?
Which Qt version and FFmpeg version exactly do you have?
Which video output and video decoder do you use?

---

Could you try to crash QMPlay2 from my DMG?

---

So far I cannot reproduce any crash on macOS (with bit-perfect). I have only on VirtualBox (44.1 kHz only available), because my HDD with macOS was corrupted by mistake. I'll install it again soon.

**Comment by RJVB** [comment]
> Which OS version do you have? Doesn't it depend on compiler and compiler tools?

I have 10.9 . Yes, it depends on the compiler, but the build scripts I use (from MacPorts) only activate sanitise support on 10.10 and up.

> Which Qt version and FFmpeg version exactly do you have?

5.9.3, FFmpeg 3.4.1

> Which video output and video decoder do you use?

OpenGL2 and the FFmpeg decoder.

> Could you try to crash QMPlay2 from my DMG?

I can try ...

Remember, I also cannot usually reproduce this crash at will.

**Comment by zaps166** [comment]
> but the build scripts I use (from MacPorts) only activate sanitise support on 10.10 and up.

I don't know which build scripts. Try with this: 3e8a45f1e30de68dfe3ae8b11368311684c88430. Define `-DUSE_ASAN=ON` and/or `-DUSE_UBSAN=ON` in cmake command and recompile.

**Comment by RJVB** [comment]
> > but the build scripts I use (from MacPorts) only activate sanitise support on 10.10 and up.
> 
> I don't know which build scripts.

The build scripts for building Clang. Apparently the libraries required for -fsanitize=X cannot be built on OS X 10.9 .
I don't know why that is, but the MacPorts packages for Clang are maintained by an Apple employee who is involved with Clang development. Clang is too expensive to build to see if he's wrong in that assessment.

**Comment by zaps166** [comment]
The best you can do is to display in logger memory address of `VideoFiltersThr` (in constructor and destructor) and compare the address after crash with address shown by debugger (if it is not null it can be also invalid). Also use global variable for `VideoFiltersThr` address, then you can try to display in logger its address when QMPlay2 crashes in signal handler.
If crash happens before destructing object and address will differ - it is a buffer overflow somewhere.

**Comment by zaps166** [comment]
Wow, I have this crash on Linux 😃

**Comment by zaps166** [comment]
@RJVB Thanks for debugging this :)

**Comment by RJVB** [comment]
You're welcome!

How did you trigger this on Linux, and figure out the cause? With the sanitise options?

**Comment by zaps166** [comment]
I tried to do this (https://github.com/zaps166/QMPlay2/issues/137#issuecomment-363942729) and adding two `qDebug()` into `VideoFiltersThr` constructor causes a delay and QMPlay2 crashed always. It uses uninitialized value from second thread (`VideoThr`) and the first thread was still in `VideoFiltersThr` constructor, so variable `filtersThr` wasn't initialized yet. Reason: thread was started too early, in incorrect constructor - more details in patch and commit description.

**Comment by RJVB** [comment]
Ah, yes, I remember also thinking of a timing issue at some point, but never thought to suggest the hypothesis.

This certainly explains why I only saw the issue just after startup when I had been doing other stuff (= the QMPlay2 and/or Qt plugins not in disk cache) and may be why my fix worked (assuming that passing by pointer instead of value is measurably slower).
